BDArmory/Control/BDGenericAIBase.cs

```diff
@@ -607,18 +597,27 @@ protected virtual void UpdateWaypoint()
                 UpdateWaypoint(); // Call ourselves again for the new waypoint to follow.
                 //Modify AI maxSpeed if the gate we just pased has a speed limit
                 float mSpeed = WaypointCourses.CourseLocations[waypointCourseIndex].waypoints[activeWaypointIndex].maxSpeed;
-                var pilotAI = VesselModuleRegistry.GetModule<BDModulePilotAI>(vessel); // Get the pilot AI if the vessel has one.
-                var surfaceAI = VesselModuleRegistry.GetModule<BDModuleSurfaceAI>(vessel); // Get the surface AI if the vessel has one.
-                var vtolAI = VesselModuleRegistry.GetModule<BDModuleVTOLAI>(vessel); // Get the VTOL AI if the vessel has one.
-                var orbitalAI = VesselModuleRegistry.GetModule<BDModuleOrbitalAI>(vessel); // Get the Orbital AI if the vessel has one.
-                if (pilotAI != null)
-                {
-                    pilotAI.maxSpeed = mSpeed > 0 ? mSpeed : originalMaxSpeed;
-                    pilotAI.OnMaxSpeedChanged();
-                }
-                if (surfaceAI != null) surfaceAI.MaxSpeed = mSpeed > 0 ? mSpeed : originalMaxSpeed;
-                //if (vtolAI != null) vtolAI.MaxSpeed = mSpeed > 0 ? mSpeed : originalMaxSpeed; //VTOL AI really needs expanding to actually include behavior beyond stationary hovering; could do WPs, but only if it has a fixed horizontal thrusters under IndependantThrottle...
-                //if (orbitalAI != null) orbitalAI.ManeuverSpeed = mSpeed > 0 ? mSpeed : originalMaxSpeed; //Don' think WPs would work, period, with an orbital reference frame?
+                var ai = vessel.ActiveController().AI;
+                if (ai != null) switch (ai.aiType)
+                    {
+                        case AIType.PilotAI:
+                            var pilotAI = ai as BDModulePilotAI;
+                            pilotAI.maxSpeed = mSpeed > 0 ? mSpeed : originalMaxSpeed;
+                            pilotAI.OnMaxSpeedChanged();
+                            break;
+                        case AIType.SurfaceAI:
+                            var surfaceAI = ai as BDModuleSurfaceAI;
+                            surfaceAI.MaxSpeed = mSpeed > 0 ? mSpeed : originalMaxSpeed;
+                            break;
+                        case AIType.VTOLAI:
+                            // var vtolAI = ai as BDModuleVTOLAI;
+                            // vtolAI.MaxSpeed = mSpeed > 0 ? mSpeed : originalMaxSpeed; //VTOL AI really needs expanding to actually include behavior beyond stationary hovering; could do WPs, but only if it has a fixed horizontal thrusters under IndependantThrottle...
```

Comment:
VTOL AI these days does have Waypoint following support, so this could be uncommented